### PR TITLE
initial check-in

### PR DESCRIPTION
**Pull request by zch401**

### CODE REVIEW

**Comment by tjiang1898** [comment]
Empty .gitignore file. Please add content to ignore .class and other none code file (see https://gist.github.com/octocat/9257657 for examples)

File: proj1a/LinkedListDeque.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    public synchronized T removeFirst() {
+        //if no first item
+        if (isEmpty()) {
+            return null;
+        }
+
+        T result = sentinel.next.item;
+        LLDNode target = sentinel.next.next;
+        sentinel.next = target;
+        size = size - 1;
```
Comment (tjiang1898) [inline proj1a/LinkedListDeque.java]: What about the target.prev that still points to the removed node? Before fixing it, could you add a unit test that would find this bug?

File: proj1a/LinkedListDeque.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    @Override
+    public synchronized T removeLast() {
+        //if no last item
+        if (isEmpty()) {
+            return null;
+        }
+
+        T result = sentinel.prev.item;
+        LLDNode target = sentinel.prev.prev;
+        sentinel.prev = target;
```
Comment (tjiang1898) [inline proj1a/LinkedListDeque.java]: Same bug above

File: proj1a/ArrayDeque.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        nextLast = size;
+        items = temp;
+    }
+
+    /**
+     * add item to the front of the deque
+     * @param item: item to add
+     */
+    @Override
+    public void addFirst(T item) {
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java]: Add `synchronized` to make it thread-safe

File: proj1a/ArrayDeque.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    @Override
+    public T removeFirst() {
+        /**if first item does not exist*/
+        if (size == 0) {
+            return null;
+        }
+
+        //R: array usage
+        double R;
+        R = size / items.length;
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java]: change it to `double R = size / items.length;` for better readability

File: proj1a/ArrayDeque.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        int old_size = size;
+        int first = nextF(nextFirst);
+
+        for (int i = 0; i < old_size ; i++ ) {
+            temp[i] = items[first];
+            first = nextF(first);
+        }
+        nextFirst = capacity - 1;
+        nextLast = size;
+        items = temp;
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java]: Would downsizing work if items are in the second half of the array? Before fixing it, could you write a unit test that would catch this bug?

File: proj1a/LinkedListDeque.java (hunk 1 above)
Comment (zch401) [reply proj1a/LinkedListDeque.java]: Bug fixed, unit test added in ```LinkedListTest.java```

File: proj1a/LinkedListDeque.java (hunk 2 above)
Comment (zch401) [reply proj1a/LinkedListDeque.java]: Bug fixed

File: proj1a/ArrayDeque.java (hunk 3 above)
Comment (zch401) [reply proj1a/ArrayDeque.java]: ``` synchorized``` added

File: proj1a/ArrayDeque.java (hunk 4 above)
Comment (zch401) [reply proj1a/ArrayDeque.java]: fixed

File: proj1a/ArrayDeque.java (hunk 5 above)
Comment (zch401) [reply proj1a/ArrayDeque.java]: ``` resize ``` method updated, test added in ```ArrayDequeTest.java ```

File: proj1a/ArrayDeque.java [hunk 6]
```diff
@@ -45,12 +40,13 @@ private int nextL(int i) {
     private void resize(int capacity) {
         T[] temp = (T[]) new Object[capacity];
         int old_size = size;
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java]: `old_size` is an unnecessary variable that made me think there was some logic depending on it. Remove it when you get a chance.

File: proj1a/ArrayDeque.java [hunk 7]
```diff
@@ -45,12 +40,13 @@ private int nextL(int i) {
...
-        int first = nextF(nextFirst);
+        int first = nextL(nextFirst);
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java]: Good fix

**Review by tjiang1898 — APPROVED** [review]

**Comment by tjiang1898** [comment]
The pull request looks good now. Please grant me permission to merge the pull request to the master branch.

**Comment by tjiang1898** [comment]
Let's have a meeting tomorrow to discuss following:
- Restructuring the project into a maven project
- Turn the unit test into JUnit tests
- Run code coverage to see how good the unit tests are.
- Turn on the CI (continuous integration) via [CircleCI ](https://circleci.com/)to trigger test automatically upon commits.

**Comment by tjiang1898** [comment]
Could you give me write access to this repo? Go to project settings tab to add me.

**Comment by zch401** [comment]
> 
> 
> Could you give me write access to this repo? Go to project settings tab to add me.

I have added you as collaborator.

File: proj1a/ArrayDeque.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+    /**
+     * create a deep copy
+     * @param other: the ArrayDeque to copy
+     */
+    public ArrayDeque(ArrayDeque other) {
+        items = (T[]) new Object[other.items.length];
+        size = 0;
+        nextFirst = 4;
+        nextLast = 5;
```
Comment (tjiang1898) [inline proj1a/ArrayDeque.java]: need to copy `cap`

**Comment by zch401** [comment]
> Let's have a meeting tomorrow to discuss following:
> 
> * Restructuring the project into a maven project
> * Turn the unit test into JUnit tests
> * Run code coverage to see how good the unit tests are.
> * Turn on the CI (continuous integration) via [CircleCI ](https://circleci.com/)to trigger test automatically upon commits.

Maven file structure added, working on CircleCI

**Comment by zch401** [comment]
> Let's have a meeting tomorrow to discuss following:
> 
> * Restructuring the project into a maven project
> * Turn the unit test into JUnit tests
> * Run code coverage to see how good the unit tests are.
> * Turn on the CI (continuous integration) via [CircleCI ](https://circleci.com/)to trigger test automatically upon commits.

CircleCI config file updated

**Comment by tjiang1898** [comment]
Excellent code coverage ... Good job!

Please merge this code into the master

**Pull request merged**